<h3>RSA</h3>
<p>El m&eacute;todo est&aacute; descrito en la hoja que se ha repartido en clase, y en el PDF que acompa&ntilde;a esta hoja.</p>
<p>Como queremos encriptar palabras de $5$ letras de un alfabeto de $26$ necesitamos una clave p&uacute;blica $n$ mayor que el n&uacute;mero de tales palabras $26^5$ pero menor que $26^6$.&nbsp;</p>

In [1]:
p = next_prime(26^3+5357)

In [2]:
q = next_prime(26^2+10635);p;q;p*q;26^5;26^6

308915776

In [3]:
n = p*q; phi = (p-1)*(q-1)

In [4]:
EX = xgcd(17,phi);EX

(1, 15267281, -1)

In [5]:
EX[1]*17-phi

1

<p>La clave p&uacute;blica del usuario es el par $(259578029,17)$,&nbsp; mientras que su clave privada es $15267281$.</p>

In [6]:
alfb = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

In [7]:
L_alfb = list(alfb);print L_alfb

['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']


In [8]:
C = "TONTO"

In [9]:
def ord2(c):
    return L_alfb.index(c)

In [10]:
def chr2(n):
    return L_alfb[n]

In [11]:
COD = [ord2(c) for c in C];COD

[19, 14, 13, 19, 14]

In [12]:
m = ZZ(COD,26); m

6740779

In [13]:
ENC = m^17%n;ENC

246929463

In [14]:
DIG = ENC.digits(base=26);DIG

[1, 7, 6, 9, 20, 20]

In [15]:
from string import *
M_ENC = join([chr2(item) for item in DIG],sep="");M_ENC

'BHGJUU'

In [16]:
M_ENC2 = [ord2(c) for c in M_ENC];M_ENC2

[1, 7, 6, 9, 20, 20]

In [17]:
m2 = ZZ(M_ENC2,26); m2

246929463

<p>Como debe ser, $m2$ es el mismo que $ENC$.</p>

In [18]:
time DESENC = m2^EX[1]%n;DESENC

CPU times: user 3.12 s, sys: 88 ms, total: 3.2 s
Wall time: 3.2 s


In [19]:
DESENCDIG = DESENC.digits(base=26);DESENCDIG

[19, 14, 13, 19, 14]

In [20]:
MENSAJE = join([chr2(item) for item in DESENCDIG],sep="");MENSAJE

'TONTO'